storage/src/main/java/io/atomix/storage/journal/JournalSegmentReader.java

```diff
@@ -136,7 +136,7 @@ private void readNext() {
     Indexed cachedEntry = cache.get(index);
     if (cachedEntry != null) {
       this.nextEntry = cachedEntry;
-      buffer.skip(cachedEntry.size() + Bytes.INTEGER + Bytes.INTEGER);
+      buffer.skip(memory.position() + cachedEntry.size() + Bytes.INTEGER + Bytes.INTEGER);
```

Comment:
@kuujo is it feasible to introduce some unit tests to assert the correct behaviour?

Reply:
I will be adding unit tests before this is merged, but some of these cases are pretty complicated

Reply:
Actually, just setting the cache size to `1` reproduces this in tests